cmd/contour/serve.go

```diff
@@ -352,6 +383,92 @@ func doServe(log logrus.FieldLogger, ctx *serveContext) error {
 	}
 	g.Add(debugsvc.Start)
 
+	// Leader Election starts here.
+
+	leaderOK := make(chan struct{})
+	deposed := make(chan struct{})
+	leadingDisabled := make(chan struct{})
+
+	g.Add(func(stop <-chan struct{}) error {
+		log := log.WithField("context", "leaderelect")
+		if ctx.DisableLeaderElection {
+			log.Info("Leader election disabled")
+			// if leader election is disabled, signal the gRPC goroutine
+			// to start serving and block this goroutine forever.
+			close(leaderOK)
+			<-leadingDisabled
+		}
+
+		log.Info("Started leader election")
+		log.Infof("Leader election params: %#v", ctx.LeaderElectionConfig)
+
+		// Generate the event recorder to send election events to the logs.
+		// Set up the event bits
+		eventBroadcaster := record.NewBroadcaster()
+		eventBroadcaster.StartLogging(log.Infof)
+		// Broadcast election events to the config map
+		eventBroadcaster.StartRecordingToSink(&clientcorev1.EventSinkImpl{Interface: clientcorev1.New(client.CoreV1().RESTClient()).Events("")})
+		eventsScheme := runtime.NewScheme()
+		// Need an eventsScheme
+		err := coreV1.AddToScheme(eventsScheme)
+		check(err)
+		// The recorder is what will record events about the resource lock
+		recorder := eventBroadcaster.NewRecorder(eventsScheme, coreV1.EventSource{Component: "contour"})
+
+		// Figure out the resource lock ID
+		var resourceLockID string
+		resourceLockID, isPodNameEnvSet := os.LookupEnv("POD_NAME")
+		if !isPodNameEnvSet {
+			resourceLockID = uuid.New().String()
+		}
+
+		rl, err := resourcelock.New(
+			resourcelock.ConfigMapsResourceLock,
+			ctx.LeaderElectionConfig.Namespace,
+			ctx.LeaderElectionConfig.Name,
+			client.CoreV1(),
+			coordinationClient,
+			resourcelock.ResourceLockConfig{
+				Identity:      resourceLockID,
+				EventRecorder: recorder,
+			},
+		)
+		check(err)
+
+		le, err := leaderelection.NewLeaderElector(leaderelection.LeaderElectionConfig{
+			Lock:          rl,
+			LeaseDuration: ctx.LeaderElectionConfig.LeaseDuration,
+			RenewDeadline: ctx.LeaderElectionConfig.RenewDeadline,
+			RetryPeriod:   ctx.LeaderElectionConfig.RetryPeriod,
+			Callbacks: leaderelection.LeaderCallbacks{
+				OnStartedLeading: func(ctx context.Context) {
+					log.WithFields(logrus.Fields{
+						"lock":     rl.Describe(),
+						"identity": rl.Identity(),
+					}).Info("started leading")
+					close(leaderOK)
+				},
+				OnStoppedLeading: func() {
+					// The context being cancelled will trigger a handler that will
+					// deal with being deposed.
+					close(deposed)
+				},
+			},
+		})
+		check(err)
+		// Running the leader election requires a context.Context, but we're
+		// not using them, so we'll just pass in an empty one.
+		le.Run(context.Background())
```

Comment:
What causes le.Run to exit? I'm guessing its the context being canceled. group.Run won't return until this function returns so we might need to make sure the context is canceled.

Reply:
I've had a crack at this, PTAL.